### DISCUSSION TITLE
Simplify DAX statement

### DISCUSSION THREAD

**Discussion by DrYSG** [DaxStudio/DaxStudio#1274] [Q&A] [answered]
In order to get data from the MSFT Teams Call Quality Data warehouse (OLAP). https://learn.microsoft.com/en-us/microsoftteams/dimensions-and-measures-available-in-call-quality-dashboard

There are a bunch of restrictions:

- The PBI connector only works in desktop mode, not in the cloud
- It makes a DirectQuery Connection
- It cannot handle anything other than the simplest DAX (only SUM, MAX) no MEAN, AVG, etc.
- It chops the result at 10K rows (so Ineed to find a way to do multiple EVALAUTES)
- I have something working in DAX studio, but DAX allows only one DEFINE for Multiple EVALAUTES, is there a way to factor out some of the common terms with VAR or MEASURE so that I can trim the EVALAUTE just to deal with CQD[End Time]

```

EVALUATE
    CALCULATETABLE(
        SUMMARIZECOLUMNS(
            CQD[Second Network Name],
            CQD[Second Subnet],
            CQD[Date],
            CQD[Day Of Week],
            CQD[Hour],
            CQD[Total Call Count],
            CQD[Avg Network Jitter Max],
            CQD[Audio Poor Percentage],
            CQD[Video Poor Percentage],
            CQD[Avg Round Trip Max]
        ),
        CQD[Stream Direction] = "First-to-Second",
        CQD[CDR Response Reason] = "OK",
        CQD[First UserType] = "Server",
        CQD[Second Inside Corp] = "Inside",
        CQD[Second Network Connection Detail] = "Wired",
        CQD[Test Call Type] = "NonTest",
        CQD[Session Type] = "Conf",
        CQD[Is Server Pair] IN {
            "Client : Server",
            " : Client"
        },
        CQD[Second UserType] IN {
            "User",
            "Anonymous"
        },
        NOT (CQD[Packet Utilization] IN {
            "(Blank)",
            "062: [0 - 0]",
            "064: [1 - 2)"
        }),
        AND(
            CQD[End Time] >= dt"2024-07-17T00:00:00",
            CQD[End Time] < dt"2024-07-17T18:00:00"    
        )
    )
```

**Comment by dgosbell** [comment, marked as answer]
> The PBI connector only works in desktop mode, not in the cloud

This does not sound right, there should be able to run anything in the service that you can build in Desktop. For some connectors or data sources behind network boundaries you may need to provision an on-prem data gateway.

> It makes a DirectQuery Connection
> It cannot handle anything other than the simplest DAX (only SUM, MAX) no MEAN, AVG, etc.

There are restrictions on DirectQuery models, there are a small set of DAX functions that are not supported in DQ  mode so this is probably expected.

> It chops the result at 10K rows (so Ineed to find a way to do multiple EVALAUTES)

I'm not sure what mechanism you are using that is limiting to 10K rows, some of the APIs have limits, but most are larger than 10K.

> I have something working in DAX studio, but DAX allows only one DEFINE for Multiple EVALAUTES, is there a way to factor out some of the common terms with VAR or MEASURE so that I can trim the EVALAUTE just to deal with CQD[End Time]

You can't really do this well since the filter needs to be embedded into the query. I don't think using variables would be a good idea as that would force materialization of the intermediate results.

**Comment by DrYSG** [comment]
You answered my question at the end. Thank you very much. Let me not leave you hanging with your questions.

About Desktop only see:

https://learn.microsoft.com/en-us/microsoftteams/cqd-power-bi-connector

RE: as the 10K limit, that is part of how MSFT trottles the usage. I confirmed with tests, and with the MSFT SME who knows all about that OLAP warehouse (btw, i is 120M rows of data, 901 columns according to the explorer in DAX Studio.


> Customers in Government cloud environments (GCC, GCC-H, DoD) will need to [install the connector](https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-custom-connectors) with an [on-premises data gateway](https://learn.microsoft.com/en-us/power-bi/connect-data/service-gateway-onprem) for compatibility with Power BI Service. Without an on-premises gateway, Government cloud customers will only be able to use the Microsoft Call Quality connector in Power BI Desktop.

**Comment by DrYSG** [comment]
@dgosbell 

Perhaps I did not make it clear what I wanted when I said "simplify" All I wanted was a way to move the common filters to a DEFINE, and leave the time silica filter in the EVALUATE.

Someone on SO helped me, so I thought I would post the solution here in case others want:


```
DEFINE
	VAR CommonFilters =
	FILTER(
		CQD,
		CQD[Stream Direction] = "First-to-Second" &&
		CQD[CDR Response Reason] = "OK" &&
		CQD[CDR Record Available] = "True" &&
		CQD[Call Setup Failure Reason] = "NotMediaFailure" &&
		CQD[Test Call Type] = "NonTest" &&
		NOT (ISBLANK(CQD[Second Network Name])) &&
		CQD[Is Server Pair] IN {
			"Client : Server",
			"Client : Client",
			" : Client"
		} &&
		CQD[First UserType] IN {
			"Server",
			"User",
			"Application"
		} &&
		CQD[Second UserType] IN {
			"User",
			"Anonymous"
		} &&
		NOT (CQD[Packet Utilization] IN {
			"(Blank)",
			"062: [0 - 0]",
			"064: [1 - 2)"
		})
	)
EVALUATE
	CALCULATETABLE(
		SUMMARIZECOLUMNS(
			CQD[Second Network Name],
			CQD[Second Subnet],
			CQD[Date],
			CQD[Day Of Week],
			CQD[Hour],
			CQD[Session Type],
			CQD[Transport],
			CQD[Second Network Connection Detail],
			CQD[Total Call Count],
			CQD[P50 Jitter],
			CQD[P50 Latency],
			CQD[P50 Packet Loss Rate],
			CQD[Avg Network Jitter Max],
			CQD[Audio Poor Percentage],
			CQD[Video Poor Percentage],
			CQD[Audio Poor Call Stream Count],
			CQD[Video Poor Stream Count],
			CQD[Avg Jitter Max],
			CQD[Avg Packet Loss Rate Max],
			CQD[Avg Call Duration],
			CQD[Avg Round Trip Max]
		),
		FILTER(
			CommonFilters,
			CQD[End Time] >= dt"2024-07-17T00:00:00" &&
			CQD[End Time] < dt"2024-07-17T17:00:00"
		)
	)
```

**Reply by marcosqlbi** [reply]
The filter written that way could be very inefficient. In DAX, it's better to filter individual columns in CALCULATE/CALCULATETABLE. By filtering the whole table (CQD), you may create a larger materialization that slows down performance.

**Comment by DrYSG** [comment]
@marcosqlbi  I really appreciate the insights of a SME in DAX! I am still a newbie in DAX. Can you help me with a example of how to move the filters to the SUMMERIZECOLUMNS, I could not figure out how to do that. 

Also, I am not sure there is any way to get around applying this to a "whole table (CQD)". Realize that this is an OLAP cube stored in what I think is a MSFT SSAS server. So you are not going to get a performance boost that you would typically by having multiple index columns in a traditional Relational DB.

But if you give me the DAX tip, I would love to do an A/B comparison.

**Reply by marcosqlbi** [reply]
I don't know the details - if it works for your use case (I just realized you use DIrectQuery, which makes any optimization consideration meaningless, as DirectQuery is extremely slow by itself), then it's fine.
You can find more examples about SUMMARIZECOLUMNS on https://dax.guide, but I suggest that you do not learn DAX function by function, as there are many concepts that require an understanding of the conceptual model of row context and filter context, which are unique to DAX and different than any other language.

**Comment by DrYSG** [comment]
@marcosqlbi Yep, MSFT Call Quality OLAP Warehouse only accepts DirectQuery, anything but the simplest DAX does not work (no aggregagates other than SUM, MIN, MAX), etc. I am using DAX Studio to suck the data out and pull it into a proper OneLake. (well for now, just a VertiPAQ store).
